repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -245,4 +263,24 @@ class Session(
     // Set the job group
     executeCode(statementId, cmd)
   }
+
+  private def cleanOldStatements(): Unit = {
+    if (_statements.size > numRetainedStatements) {
+      // Remove 10% of existing statements to avoid frequently calling this method when reaching
+      // threshold.
+      val toRemove = (numRetainedStatements * 0.1).toInt + 1
+      val stmtIdToRemove = math.max(newStatementId.get() - _statements.size, 0)
+      (stmtIdToRemove until stmtIdToRemove + toRemove).foreach { id =>
+        _statements.get(id).foreach { st =>
```

Comment:
My concern is that we should add explicit lock in many different places, this makes the current code hard to read. My original purpose is not high concurrency, sorry about the confuse, let me try to replace with {{LinkedHashMap}}